File: lib/java/opentoken/src/main/java/com/truveta/opentoken/tokens/TokenRegistry.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokens;

import com.truveta.opentoken.attributes.AttributeExpression;
import java.util.ServiceLoader;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TokenRegistry {

    private TokenRegistry() {

    }

    public static Map<String, List<AttributeExpression>> loadAllTokens() {
        Map<String, List<AttributeExpression>> tokensMap = new HashMap<>();
        ServiceLoader<Token> loader = ServiceLoader.load(Token.class);
        for (Token token : loader) {
            tokensMap.put(token.getIdentifier(), token.getDefinition());
        }
        return tokensMap;
    }
}

File: lib/python/opentoken/src/main/opentoken/tokens/token_registry.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import pkgutil
import importlib
import pathlib
from typing import Dict, List
from opentoken.attributes.attribute_expression import AttributeExpression
from opentoken.tokens.token import Token


class TokenRegistry:
    @staticmethod
    def load_all_tokens() -> Dict[str, List[AttributeExpression]]:
        definitions: Dict[str, List[AttributeExpression]] = {}

        # package name for import
        package = "opentoken.tokens.definitions"

        # real filesystem path (relative to this file)
        package_path = str(pathlib.Path(__file__).parent / "definitions")

        # iterate over modules in the definitions package
        for _, modname, _ in pkgutil.iter_modules([package_path]):
            module = importlib.import_module(f"{package}.{modname}")

            # scan for Token subclasses
            for obj in module.__dict__.values():
                if isinstance(obj, type) and issubclass(obj, Token) and obj is not Token:
                    token = obj()
                    definitions[token.get_identifier()] = token.get_definition()

        return definitions
